// @adviser/cement/cf/index.js
import {
  WebFileService
} from "../chunk-H6MLKWIZ.js";
import {
  BaseSysAbstraction,
  CFEnvActions,
  WrapperSysAbstraction,
  envFactory
} from "../chunk-23MCJYE6.js";
import {
  TxtEnDecoderSingleton
} from "../chunk-LXTKPZ6M.js";

// src/cf/cf-sys-abstraction.ts
var CFSystemService = class {
  Env() {
    return envFactory();
  }
  Args() {
    throw new Error("Args-Method not implemented.");
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  OnExit(hdl) {
    throw new Error("OnExit-Method not implemented.");
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  Exit(code) {
    throw new Error("Exit-Method not implemented.");
  }
};
var my = void 0;
function CFSysAbstraction(param) {
  if (!my) {
    my = new BaseSysAbstraction({
      TxtEnDecoder: (param == null ? void 0 : param.TxtEnDecoder) || TxtEnDecoderSingleton(),
      FileSystem: new WebFileService(),
      SystemService: new CFSystemService()
    });
  }
  return new WrapperSysAbstraction(my, param);
}
export {
  CFEnvActions,
  CFSysAbstraction,
  CFSystemService
};
