// @adviser/cement/ts/src/deno/deno-file-service.js
import * as path from "node:path";
import { TxtEnDecoderSingleton } from "../txt-en-decoder.js";
const Deno = globalThis.Deno;
export class DenoFileService {
    constructor(baseDir = Deno.cwd(), txtEnde = TxtEnDecoderSingleton()) {
        this.baseDir = this.abs(baseDir);
        this.txtEnde = txtEnde;
    }
    async readFileString(fname) {
        return this.txtEnde.decode(await Deno.readFile(fname));
    }
    dirname(fname) {
        return path.dirname(fname);
    }
    basename(fname) {
        return path.basename(fname);
    }
    join(...paths) {
        return path.join(...paths);
    }
    relative(from, to) {
        if (to === undefined) {
            to = from;
            from = Deno.cwd();
        }
        const ret = path.relative(from, to);
        return ret;
    }
    abs(fname) {
        if (path.isAbsolute(fname)) {
            return fname;
        }
        else {
            const cwd = Deno.cwd();
            return path.resolve(cwd, fname);
        }
    }
    isAbsolute(fname) {
        return path.isAbsolute(fname);
    }
    async writeFileString(fname, content, ende = TxtEnDecoderSingleton()) {
        const o = await this.create(fname);
        const wr = o.stream.getWriter();
        await wr.write(ende.encode(content));
        await wr.close();
    }
    async create(fname) {
        let oName = fname;
        if (!path.isAbsolute(fname)) {
            oName = this.abs(fname);
        }
        const base = path.dirname(oName);
        await Deno.mkdir(base, { recursive: true });
        const out = await Deno.open(oName, {
            write: true,
            create: true,
            truncate: true,
        });
        return {
            name: oName,
            stream: out,
        };
    }
}
