// @fireproof/core/web/index.js
// src/runtime/gateways/indexdb/web/gateway-impl.ts
import { openDB } from "idb";
import { exception2Result, KeyedResolvOnce, Result } from "@adviser/cement";

// src/runtime/gateways/indexdb/version.ts
var INDEXDB_VERSION = "v0.19-indexdb";

// src/runtime/gateways/indexdb/web/gateway-impl.ts
import { ensureLogger, exceptionWrapper, getKey, getStore, NotFoundError } from "@fireproof/core";
function ensureVersion(url) {
  return url.build().defParam("version", INDEXDB_VERSION).URI();
}
function sanitzeKey(key) {
  if (key.length === 1) {
    key = key[0];
  }
  return key;
}
var onceConn = new KeyedResolvOnce();
async function connectIdb(url, sthis) {
  const dbName = getIndexDBName(url, sthis);
  const once = await onceConn.get(dbName.fullDb).once(async () => {
    const db = await openDB(dbName.fullDb, 1, {
      upgrade(db2) {
        ["version", "data", "wal", "meta", "idx.data", "idx.wal", "idx.meta"].map((store) => {
          db2.createObjectStore(store, {
            autoIncrement: false
          });
        });
      }
    });
    const found = await db.get("version", "version");
    const version = ensureVersion(url).getParam("version");
    if (!found) {
      await db.put("version", { version }, "version");
    } else if (found.version !== version) {
      sthis.logger.Warn().Str("url", url.toString()).Str("version", version).Str("found", found.version).Msg("version mismatch");
    }
    return { db, dbName, version, url };
  });
  return {
    ...once,
    url: url.build().setParam("version", once.version).URI()
  };
}
function joinDBName(...names) {
  return names.map((i) => i.replace(/^[^a-zA-Z0-9]+/g, "").replace(/[^a-zA-Z0-9-]+/g, "_")).filter((i) => i.length).join(".");
}
function getIndexDBName(iurl, sthis) {
  const url = ensureVersion(iurl);
  const fullDb = url.pathname.replace(/^\/+/, "").replace(/\?.*$/, "");
  const dbName = url.getParam("name");
  if (!dbName) throw sthis.logger.Error().Str("url", url.toString()).Msg(`name not found`).AsError();
  const result = joinDBName(fullDb, dbName);
  const objStore = getStore(url, sthis, joinDBName).name;
  const connectionKey = [result, objStore].join(":");
  return {
    fullDb: result,
    objStore,
    connectionKey,
    dbName
  };
}
var IndexDBGateway = class {
  constructor(sthis) {
    this.logger = ensureLogger(sthis, "IndexDBGateway");
    this.sthis = sthis;
  }
  // _db: IDBPDatabase<unknown> = {} as IDBPDatabase<unknown>;
  async start(baseURL) {
    return exception2Result(async () => {
      this.logger.Debug().Url(baseURL).Msg("starting");
      await this.sthis.start();
      const ic = await connectIdb(baseURL, this.sthis);
      this.logger.Debug().Url(ic.url).Msg("started");
      return ic.url;
    });
  }
  async close() {
    return Result.Ok(void 0);
  }
  async destroy(baseUrl) {
    return exception2Result(async () => {
      const type = getStore(baseUrl, this.sthis, joinDBName).name;
      const idb = await connectIdb(baseUrl, this.sthis);
      const trans = idb.db.transaction(type, "readwrite");
      const object_store = trans.objectStore(type);
      const toDelete = [];
      for (let cursor = await object_store.openCursor(); cursor; cursor = await cursor.continue()) {
        toDelete.push(cursor.primaryKey);
      }
      for (const key of toDelete) {
        await trans.db.delete(type, key);
      }
      await trans.done;
    });
  }
  buildUrl(baseUrl, key) {
    return Promise.resolve(Result.Ok(baseUrl.build().setParam("key", key).URI()));
  }
  async get(url) {
    return exceptionWrapper(async () => {
      const key = getKey(url, this.logger);
      const store = getStore(url, this.sthis, joinDBName).name;
      this.logger.Debug().Url(url).Str("key", key).Str("store", store).Msg("getting");
      const { db } = await connectIdb(url, this.sthis);
      const tx = db.transaction([store], "readonly");
      const bytes = await tx.objectStore(store).get(sanitzeKey(key));
      await tx.done;
      if (!bytes) {
        return Result.Err(new NotFoundError(`missing ${key}`));
      }
      return Result.Ok(bytes);
    });
  }
  async put(url, value) {
    return exception2Result(async () => {
      const key = getKey(url, this.logger);
      const store = getStore(url, this.sthis, joinDBName).name;
      this.logger.Debug().Url(url).Str("key", key).Str("store", store).Msg("putting");
      const { db } = await connectIdb(url, this.sthis);
      const tx = db.transaction([store], "readwrite");
      await tx.objectStore(store).put(value, sanitzeKey(key));
      await tx.done;
    });
  }
  async delete(url) {
    return exception2Result(async () => {
      const key = getKey(url, this.logger);
      const store = getStore(url, this.sthis, joinDBName).name;
      this.logger.Debug().Url(url).Str("key", key).Str("store", store).Msg("deleting");
      const { db } = await connectIdb(url, this.sthis);
      const tx = db.transaction([store], "readwrite");
      await tx.objectStore(store).delete(sanitzeKey(key));
      await tx.done;
      return Result.Ok(void 0);
    });
  }
};
var IndexDBTestGateway = class {
  constructor(sthis) {
    this.sthis = sthis;
    this.logger = ensureLogger(sthis, "IndexDBTestStore", {});
  }
  async get(url, key) {
    const ic = await connectIdb(url, this.sthis);
    const store = getStore(ic.url, this.sthis, joinDBName).name;
    this.logger.Debug().Str("key", key).Str("store", store).Msg("getting");
    let bytes = await ic.db.get(store, sanitzeKey(key));
    this.logger.Debug().Str("key", key).Str("store", store).Int("len", bytes.length).Msg("got");
    if (typeof bytes === "string") {
      bytes = this.sthis.txt.encode(bytes);
    }
    return bytes;
  }
};

// src/runtime/gateways/indexdb/web/key-bag-indexdb.ts
import { openDB as openDB2 } from "idb";
import { ResolveOnce } from "@adviser/cement";
import { rt } from "@fireproof/core";
var KeyBagProviderIndexDB = class {
  constructor(url, sthis) {
    this._db = new ResolveOnce();
    this.sthis = sthis;
    this.logger = sthis.logger;
    this.url = url;
    this.dbName = rt.getPath(this.url, this.sthis);
  }
  async _prepare() {
    return this._db.once(async () => {
      return await openDB2(this.dbName, 1, {
        upgrade(db) {
          ["bag"].map((store) => {
            db.createObjectStore(store, {
              autoIncrement: false
            });
          });
        }
      });
    });
  }
  async get(id) {
    const db = await this._prepare();
    const tx = db.transaction(["bag"], "readonly");
    const keyItem = await tx.objectStore("bag").get(id);
    await tx.done;
    if (!keyItem) {
      return void 0;
    }
    return keyItem;
  }
  async set(id, item) {
    const db = await this._prepare();
    const tx = db.transaction(["bag"], "readwrite");
    await tx.objectStore("bag").put(item, id);
    await tx.done;
  }
};
export {
  IndexDBGateway as GatewayImpl,
  IndexDBTestGateway as GatewayTestImpl,
  IndexDBGateway,
  IndexDBTestGateway,
  KeyBagProviderIndexDB as KeyBagProviderImpl,
  KeyBagProviderIndexDB,
  getIndexDBName
};
